### ISSUE TITLE
Mocha somehow modifying function-level state of describe/it blocks?

### ISSUE THREAD

**Issue by jamesmontalvo3** [xOPERATIONS/maestro#104]
In trying to create the following test (`./app/model/ColumnsHandler.spec.js`), the state of a variable inside a function appears to be getting modified from outside. The version of `ColumnsHandler.js` this was run against can be found in 872ff88708727df73e80cdd67ded5ccb29389c8d, [here](https://github.com/xOPERATIONS/maestro/blob/872ff88708727df73e80cdd67ded5ccb29389c8d/app/model/ColumnsHandler.js). 

Test that's not behaving as expected:

```javascript
/* Specify environment to include mocha globals */
/* eslint-env node, mocha */

'use strict';

const assert = require('chai').assert;
const ColumnsHandler = require('./ColumnsHandler');

describe('ColumnsHandler', function() {

	function initializeTwoColumns() {
		const ch = new ColumnsHandler();
		const columnsDef = [
			{ key: 'IV', actors: '*', display: 'IV/SSRMS/MCC' }, // note "actors" is string here
			{ key: 'EV1', actors: ['EV1'], display: 'EV1' } // note "actors" is array here
		];
		ch.updateColumns(columnsDef);
		columnsDef[0].actors = ['*']; // in model, converts strings to arrays for actors
		return {
			columnsHandler: ch,
			columnsDef: columnsDef
		};
	}

	describe('getActorColumnKey()', function() {

		// this should generate a setup with two columns: IV and EV1
		const ch = initializeTwoColumns();
		console.log(ch.columnsHandler); // --> shows IV and EV1 column, no EV2 column

		// this passes, as it correctly finds that EV1 belongs in the EV1 column
		it('should say EV1 is in EV1 column', function() {
			console.log(ch.columnsHandler); // --> IV, EV1, and magically an EV2 column!?
			assert.strictEqual(ch.columnsHandler.getActorColumnKey('EV1'), 'EV1');
		});

		// this FAILS, as it thinks EV2 should be in the EV2 column, not in the IV column. But EV2
		// should be in the IV column because above _no EV2 column was created_! Somehow the state
		// of the 'ch' variable is being modified.
		it('should say EV2 is in IV column when EV2 column not yet defined', function() {
			console.log(ch.columnsHandler); // --> IV, EV1, and magically an EV2 column!?
			assert.strictEqual(ch.columnsHandler.getActorColumnKey('EV2'), 'IV');
		});

		// this passes, despite being "equivalent" to the above.
		it('and here\'s a test of ^ those two with initializeTwoColumns() inside it()', function() {
			const ch = initializeTwoColumns();
			console.log(ch.columnsHandler); // --> shows IV and EV1 column, no EV2 column
			assert.strictEqual(ch.columnsHandler.getActorColumnKey('EV1'), 'EV1');
			assert.strictEqual(ch.columnsHandler.getActorColumnKey('EV2'), 'IV');
		});
	});
});
```

I'm calling this a Mocha issue, but it's definitely possible it's a me-issue.

**Comment by cameronwp** [comment]
This line is the problem: https://github.com/xOPERATIONS/maestro/blob/mocha-weirdness/app/model/ColumnsHandler.spec.js#L98

`describe` and `it` allow you to pass callback functions to mocha. When you write tests, you're essentially registering the callbacks with mocha, which then takes responsibility for executing them. Meanwhile, all the code in between the callbacks is essentially "hoisted" to the top of the callback. In the case of multiple `it`s within a `describe`, any code outside of an `it` may be run _before_ any of the callbacks passed to the `its` will run. There's no guarantee on the ordering.

In other words, the line above is running before any of the earlier tests. Commenting it out fixes the tests (but breaks later tests that expect it to have been run).

To fix, I would explicitly initialize a new `ColumnHandler` and do whatever you need to it within each `it` callback.

**Comment by cameronwp** [comment]
It's generally best practice to stick to [hooks](https://mochajs.org/#hooks) if you want to run code outside tests.
